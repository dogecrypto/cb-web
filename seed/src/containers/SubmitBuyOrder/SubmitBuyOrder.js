import React, { Component } from 'react'
import { BigNumber } from '@0xproject/utils'
import { ellipsizeString } from '../../utils/utils.js'
import { getSymbolFromCurrency } from '../../constants/currencyConstants.js'
import strings from '../../locales/default.js'

export class SubmitBuyOrderComponent extends Component {
  constructor (props) {
    super(props)
    this.props.fetchTokenList()
    this.props.fetchExchangeRates()
    this.state = {
      inputFiatAmountEstimate: '',
      inputAmount: '',
      outputFiatAmountEstimate: '',
      outputAmount: ''
    }
  }
  handleSubmit = (e) => {
    const { submitOrder, inputCurrencyCode, outputCurrencyCode } = this.props
    const { inputAmount, outputAmount } = this.state
    console.log('handling submit: ', e)
    e.preventDefault()
    submitOrder({
      inputAmount,
      outputAmount,
      inputCurrencyCode,
      outputCurrencyCode
    })
  }

  onChangeInputCurrencyCode = (event) => {
    const { updateInputCurrencyCode } = this.props
    updateInputCurrencyCode(event.target.value)
  }

  onChangeInputAmount = (event) => {
    const { inputCurrencyFiatRate } = this.props
    let inputFiatAmountEstimate = ''
    const inputAmount = new BigNumber(event.target.value)
    if (inputCurrencyFiatRate) inputFiatAmountEstimate = inputAmount.mul(inputCurrencyFiatRate).toFixed(2).toString()
    this.setState({
      inputAmount: inputAmount.toString(),
      inputFiatAmountEstimate
    })
  }

  onChangeOutputAmount = (event) => {
    const { outputCurrencyFiatRate } = this.props
    let outputFiatAmountEstimate = ''
    const outputAmount = new BigNumber(event.target.value)
    if (outputCurrencyFiatRate) outputFiatAmountEstimate = outputAmount.mul(outputCurrencyFiatRate).toFixed(2).toString()
    this.setState({
      outputAmount: outputAmount.toString(),
      outputFiatAmountEstimate
    })
  }

  onChangeOutputCurrencyCode = (event) => {
    const { updateOutputCurrencyCode } = this.props
    updateOutputCurrencyCode(event.target.value)
  }

  render () {
    const {
      account,
      tokenDirectory,
      inputCurrencyCode,
      inputCurrencyFiatRate,
      outputCurrencyCode,
      outputCurrencyFiatRate,
      isoFiatCurrencyCode,
      isOrderFormProcessing,
      orderFormFeedback
    } = this.props
    const { inputFiatAmountEstimate, outputFiatAmountEstimate } = this.state
    let alertClass = ''
    const fiatCurrencySymbol = getSymbolFromCurrency(isoFiatCurrencyCode)
    const inputCurrencyFiatRateSyntax = inputCurrencyFiatRate ? `~ ${fiatCurrencySymbol} ${inputCurrencyFiatRate}` : ''
    const inputEstimate = inputFiatAmountEstimate ? `~ ${fiatCurrencySymbol} ${inputFiatAmountEstimate}` : ''
    const outputCurrencyFiatRateSyntax = outputCurrencyFiatRate ? `~ ${fiatCurrencySymbol} ${outputCurrencyFiatRate}` : ''
    const outputEstimate = outputFiatAmountEstimate ? `~ ${fiatCurrencySymbol} ${outputFiatAmountEstimate}` : ''

    if (orderFormFeedback.type) alertClass = `alert alert-${orderFormFeedback.type}`

    return (
      <div>
        <div className='card__title'>
          <h5 className='bold-text'>Choose trading pairs:</h5>
        </div>
        <form className='form form--horizontal' onSubmit={this.handleSubmit}>
          <div className='form__form-group form-row'>
            <div className='col-3 form__form-group-label-wrap'>
              <label className='form__form-group-label' htmlFor='inputCurrencyCode'>Token to Sell:</label>
            </div>
            <div className='col-7'>
              <select value={inputCurrencyCode} onChange={this.onChangeInputCurrencyCode} className='form-control'>
                <option key={null} value={null}>----</option>
                {tokenDirectory.map((token) => {
                  const tokenCode = token.symbol.split(' ')
                  return (
                    <option key={token.symbol} value={token.symbol}>{tokenCode[0]} | {ellipsizeString(token.address, 20).toLowerCase()}</option>
                  )
                })}
              </select>
            </div>
            <div className='col-2 form__form-group-label-wrap post'>
              <span className='form__form-group-label'>{inputCurrencyFiatRateSyntax}</span>
            </div>
          </div>

          <div className='form__form-group form-row'>
            <div className='col-3 form__form-group-label-wrap'>
              <label className='form__form-group-label' htmlFor='inputCurrencyAmount'>Amount to Sell:</label>
            </div>
            <div className='col-7'>
              <input className='form-control' type='number' placeholde='0' onChange={this.onChangeInputAmount} step='0.000000000000000001'/>
            </div>
            <div className='col-2 form__form-group-label-wrap post'>
              <span className='form__form-group-label'>{inputEstimate}</span>
            </div>
          </div>

          <div className='form__form-group form-row'>
            <div className='col-3 form__form-group-label-wrap'>
              <label className='form__form-group-label' htmlFor='outputCurrencyCode'>Token to Buy:</label>
            </div>
            <div className='col-7'>
              <select value={outputCurrencyCode} onChange={this.onChangeOutputCurrencyCode} className='form-control'>
                <option key={null} value={null}>----</option>
                {tokenDirectory.map((token) => {
                  const tokenCode = token.symbol.split(' ')
                  return (
                    <option key={token.symbol} value={token.symbol}>{tokenCode[0]} | {ellipsizeString(token.address, 20).toLowerCase()}</option>
                  )
                })}
              </select>
            </div>

            <div className='col-2 form__form-group-label-wrap post'>
              <span className='form__form-group-label'>{outputCurrencyFiatRateSyntax}</span>
            </div>
          </div>
          <div className='form__form-group form-row'>
            <div className='col-3 form__form-group-label-wrap'>
              <label className='form__form-group-label' htmlFor='outputCurrencyAmount'>Amount to Buy:</label>
            </div>
            <div className='col-7'>
              <input className='form-control' type='number' placeholde='0' onChange={this.onChangeOutputAmount} step='0.000000000000000001' />
            </div>
            <div className='col-2 form__form-group-label-wrap post'>
              <span className='form__form-group-label'>{outputEstimate}</span>
            </div>
          </div>
          <div className={`feedback-area alert ${alertClass} form-row`}>
            {orderFormFeedback.message && <span>{orderFormFeedback.message}</span>}
          </div>
          <div className='centered-form-buttons form-row'>
            <button className='btn btn-secondary btn-lg' style={{ width: 200 }} type='button'>{strings.clear}</button>
            <button className='btn btn-primary btn-lg' style={{ width: 200 }} disabled={!account || isOrderFormProcessing} type='submit'>{isOrderFormProcessing ? 'Processing...' : strings.submit}</button>
          </div>
        </form>
      </div>
    )
  }
}
