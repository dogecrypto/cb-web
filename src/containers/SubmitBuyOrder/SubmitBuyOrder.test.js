import React from 'react'
import renderer from 'react-test-renderer'
import { SubmitBuyOrderComponent } from './SubmitBuyOrder'

it('SubmitBuyOrder component initally renders accurately', () => {
  const fakeTokenDirectory = [
    {
      address: '0xE41d2489571d322189246DaFA5ebDe1F4699F498',
      symbol: 'ZRX',
      decimal: 18,
      type: 'default'
    },{
      address: '0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2',
      symbol: 'WETH',
      decimal: 18,
      type: 'default'
    }
  ]
  const tree = renderer  
    .create(
      <SubmitBuyOrderComponent
        account={null}
        tokenDirectory={fakeTokenDirectory}
        inputCurrencyCode={'WETH'}
        outputCurrencyCode={'ZRX'}
        inputCurrencyFiatRate={300}
        outputCurrencyFiatRate={0.7}
        isoFiatCurrencyCode={'iso:USD'}
        isOrderFormProcessing={false}
        orderFormFeedback={''}
        orderFormProcessingButtonTitle={'Submit'}
        submitOrder={() => {}}
        fetchTokenList={() => {}}
        fetchExchangeRates={() => {}}
        updateInputCurrencyCode={() => {}}
        updateOutputCurrencyCode={() => {}}   
      />
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
});