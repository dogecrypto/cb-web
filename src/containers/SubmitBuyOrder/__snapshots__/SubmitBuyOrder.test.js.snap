// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SubmitBuyOrder component initally renders accurately 1`] = `
<div>
  <div
    className="card__title"
  >
    <h5
      className="bold-text"
    >
      Choose trading pairs:
    </h5>
  </div>
  <form
    className="form form--horizontal"
    onSubmit={[Function]}
  >
    <div
      className="form__form-group form-row"
    >
      <div
        className="col-3 form__form-group-label-wrap"
      >
        <label
          className="form__form-group-label"
          htmlFor="inputCurrencyCode"
        >
          Token to Sell:
        </label>
      </div>
      <div
        className="col-7"
      >
        <select
          className="form-control"
          onChange={[Function]}
          value="WETH"
        >
          <option
            value={null}
          >
            ----
          </option>
          <option
            value="ZRX"
          >
            ZRX
             | 
            0xe41d2489...1f4699f498
          </option>
          <option
            value="WETH"
          >
            WETH
             | 
            0xc02aaa39...083c756cc2
          </option>
        </select>
      </div>
      <div
        className="col-2 form__form-group-label-wrap post"
      >
        <span
          className="form__form-group-label"
        >
          ~ $ 300
        </span>
      </div>
    </div>
    <div
      className="form__form-group form-row"
    >
      <div
        className="col-3 form__form-group-label-wrap"
      >
        <label
          className="form__form-group-label"
          htmlFor="inputCurrencyAmount"
        >
          Amount to Sell:
        </label>
      </div>
      <div
        className="col-7"
      >
        <input
          className="form-control"
          onChange={[Function]}
          placeholde="0"
          step="0.000000000000000001"
          type="number"
        />
      </div>
      <div
        className="col-2 form__form-group-label-wrap post"
      >
        <span
          className="form__form-group-label"
        >
          
        </span>
      </div>
    </div>
    <div
      className="form__form-group form-row"
    >
      <div
        className="col-3 form__form-group-label-wrap"
      >
        <label
          className="form__form-group-label"
          htmlFor="outputCurrencyCode"
        >
          Token to Buy:
        </label>
      </div>
      <div
        className="col-7"
      >
        <select
          className="form-control"
          onChange={[Function]}
          value="ZRX"
        >
          <option
            value={null}
          >
            ----
          </option>
          <option
            value="ZRX"
          >
            ZRX
             | 
            0xe41d2489...1f4699f498
          </option>
          <option
            value="WETH"
          >
            WETH
             | 
            0xc02aaa39...083c756cc2
          </option>
        </select>
      </div>
      <div
        className="col-2 form__form-group-label-wrap post"
      >
        <span
          className="form__form-group-label"
        >
          ~ $ 0.7
        </span>
      </div>
    </div>
    <div
      className="form__form-group form-row"
    >
      <div
        className="col-3 form__form-group-label-wrap"
      >
        <label
          className="form__form-group-label"
          htmlFor="outputCurrencyAmount"
        >
          Amount to Buy:
        </label>
      </div>
      <div
        className="col-7"
      >
        <input
          className="form-control"
          onChange={[Function]}
          placeholde="0"
          step="0.000000000000000001"
          type="number"
        />
      </div>
      <div
        className="col-2 form__form-group-label-wrap post"
      >
        <span
          className="form__form-group-label"
        >
          
        </span>
      </div>
    </div>
    <div
      className="feedback-area alert  form-row"
    />
    <div
      className="centered-form-buttons form-row"
    >
      <button
        className="btn btn-secondary btn-lg"
        style={
          Object {
            "width": 200,
          }
        }
        type="button"
      >
        Clear
      </button>
      <button
        className="btn btn-primary btn-lg"
        disabled={true}
        style={
          Object {
            "width": 200,
          }
        }
        type="submit"
      >
        Submit
      </button>
    </div>
  </form>
</div>
`;
